### PR TITLE
Fixing issue #5057 - ZStream collectType operator

### PR DESCRIPTION
**Pull request by remiguittaut**
Add the collectType operator to collect elements of a given subtype in ZStreams. The implementation must be defined as an extension method in an implicit class because of variance issues. The way to define the implicit class is up for discussion.

### CODE REVIEW

File: streams/shared/src/main/scala/zio/stream/ZStreamOps.scala [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package zio.stream
+
+import scala.reflect.ClassTag
+
+trait ZStreamOps {
+  implicit class Syntax[R, E, O](self: ZStream[R, E, O]) {
```
Comment (adamgfraser) [inline streams/shared/src/main/scala/zio/stream/ZStreamOps.scala:6]: You can just make this an implicit class inside the `ZStream` companion object.

Comment (remiguittaut) [reply streams/shared/src/main/scala/zio/stream/ZStreamOps.scala:6]: Ok. Wasn't sure I should touch the companion object. Will do

Comment (remiguittaut) [reply streams/shared/src/main/scala/zio/stream/ZStreamOps.scala]: Moved it to the companion object

**Review by iravid — APPROVED** [review]

**Comment by remiguittaut** [comment]
> Thank you @remiguittaut! Would you like to add this to the new ZStream encoding as well? You can PR the implementation to series/2.x.

Sorry, long vacations spent moving to a new house, I'm completely off the OSS hook :). I'll make a new PR when I have time.

**Comment by iravid** [comment]
No worries! Took care of that in #5190

**Pull request merged**